libr/io/submap.c

```diff
@@ -0,0 +1,289 @@
+#include <r_io.h>
+#include <r_types.h>
+#include <r_util.h>
+
+ut64 _map_from(RIOMap *map) {
+	return map->itv.addr;
+}
+
+ut64 _map_to(RIOMap *map) {
+	return map->itv.addr + map->itv.size - 1;
+}
+
+//in - incoming
+static bool _submap_find_cb(void *vin, void *vincoming, void *user, int *cmp_res) {
+	RIOSubMap *sm = (RIOSubMap *)vin;
+	ut64 *off = (ut64 *)vincoming;
+
+	if (off[0] < sm->from) {
+		cmp_res[0] = 1;
+		goto beach;
+	}
+	if (off[0] > sm->to) {
+		cmp_res[0] = -1;
+		goto beach;
+	}
+	cmp_res[0] = 0;
+beach:
+	return true;
+}
+
+bool _submap_free_cb(void *user, void *data, ut32 id) {
+	free(data);
+	return true;
+}
+
+//only use this, when no overlapping is guaranteed with the incoming data
+bool _submap_insert_no_overlap_cb(void *vin, void *vincoming, void *user, int *cmp_res) {
+	RIOSubMap *sm = (RIOSubMap *)vincoming;
+
+	return _submap_find_cb(vin, &sm->from, user, cmp_res);
+}
+
+R_API void r_io_submap_init(RIO *io) {
+	if (io && !io->submaps) {
+		io->submaps = r_oids_new (0, 0xfffffffe);	//we need to guarantee 1 empty slot
+	}
+}
+
+R_API void r_io_submap_fini(RIO *io) {
+	if (io && io->submaps) {
+		r_oids_foreach(io->submaps, _submap_free_cb, NULL);
+		r_oids_free(io->submaps);
+		io->submaps = NULL;
+	}
+}
+
+//existing partially overlapping submaps get adjusted,
+//so that they don't overlap with the new submap
+R_API void r_io_submap_repress(RIO *io, RIOMap *map) {	//change to bool maybe later
+	RIOSubMap *bd, *sm;
+	ut64 off;
+	ut32 od;
+
+	if (!io || !io->submaps || !map) {
+		return;
+	}
+
+	sm = R_NEW (RIOSubMap);
+	if (!sm) {
+		return;
+	}
+	sm->from = _map_from (map);
+	sm->to = _map_to (map);
+	sm->id = map->id;
+	r_io_submap_cut_out (io, sm->from, sm->to);
+	io->submaps->cmp = _submap_insert_no_overlap_cb;
+	r_oids_insert (io->submaps, sm, &sm->sid, &od, NULL);
+}
+
+//cuts out all submaps, so that there is no submap between from and to
+R_API bool r_io_submap_cut_out(RIO *io, ut64 from, ut64 to) {
+	RIOSubMap *bd, *sm;
+	ut32 od;
+	if (to < from) {
+		if (!r_io_submap_cut_out(io, from, UT64_MAX)) {
+			return false;
+		}
+		from = 0LL;
+	}
+	if (!io || !io->submaps) {
+		return false;
+	}
+	if (!io->submaps->ptop) {
+		return true;
+	}
+	io->submaps->cmp = _submap_find_cb;
+	od = r_oids_find (io->submaps, &from, NULL);
+	while ((sm = r_oids_oget(io->submaps, od))) {
+//sm->to is guaranteed to be bigger or equal to from,
+//because that's how r_oids_find works
+		if (sm->from < from) {
```

Comment:
I have told you last year this is not how intervals are supposed to be merged

Reply:
this function doesn't merge intervals